pixy/consumer.go

```diff
@@ -0,0 +1,562 @@
+			rebalanceResultCh <- err
+			return
+		}
+		groupAssignments := resolveAssignments(topicPartitions, topicSubscribers[topic])
+		topicAssignments := groupAssignments[gc.config.ClientID]
+		log.Infof("<%s> partitions assigned: topic=%s, my=%v, all=%v", cid, topic, topicAssignments, groupAssignments)
+		assignments[topic] = topicAssignments
+	}
+	// Stop partition consumers that are no more assigned to this group member.
+	var unassignedPartitions []*exclusiveConsumer
+	for topic, topicExclusiveConsumers := range gc.exclusiveConsumers {
+		topicAssignments := assignments[topic]
+		for partition, ec := range topicExclusiveConsumers {
+			if !topicAssignments[partition] {
+				delete(topicExclusiveConsumers, partition)
+				unassignedPartitions = append(unassignedPartitions, ec)
+			}
+		}
+	}
+	stopExclusiveConsumers(unassignedPartitions)
+	// Spawn consumers for just assigned partitions.
+	for topic, topicAssignments := range assignments {
+		tc := topicConsumers[topic]
+		if tc == nil {
+			continue
+		}
+		topicExclusiveConsumers := gc.exclusiveConsumers[topic]
+		if topicExclusiveConsumers == nil {
+			topicExclusiveConsumers = make(map[int32]*exclusiveConsumer, len(topicAssignments))
+			gc.exclusiveConsumers[topic] = topicExclusiveConsumers
+		}
+		for partition := range topicAssignments {
+			if _, ok := topicExclusiveConsumers[partition]; !ok {
+				ec := gc.spawnExclusiveConsumer(tc, partition)
+				topicExclusiveConsumers[partition] = ec
+			}
+		}
+	}
+	// Notify the caller that rebalancing has completed successfully.
+	rebalanceResultCh <- nil
+	return
+}
+
+func stopExclusiveConsumers(exclusiveConsumers []*exclusiveConsumer) {
+	wg := sync.WaitGroup{}
+	for _, ec := range exclusiveConsumers {
+		spawn(&wg, ec.stop)
+	}
+	wg.Wait()
+}
+
+func listTopics(topicConsumers map[string]*topicConsumer) []string {
+	topics := make([]string, 0, len(topicConsumers))
+	for topic := range topicConsumers {
+		topics = append(topics, topic)
+	}
+	return topics
+}
+
+func resolveAssignments(partitions []int32, members []string) map[string]map[int32]bool {
+	partitionCount := len(partitions)
+	memberCount := len(members)
+	if partitionCount == 0 || memberCount == 0 {
+		return nil
+	}
+	sort.Sort(Int32Slice(partitions))
+	sort.Sort(sort.StringSlice(members))
+
+	assignments := make(map[string]map[int32]bool, memberCount)
+	partitionsPerMember := partitionCount / memberCount
+	extra := partitionCount - memberCount*partitionsPerMember
+
+	begin := 0
+	for _, memberID := range members {
+		end := begin + partitionsPerMember
+		if extra != 0 {
+			end++
+			extra--
+		}
+		for _, partition := range partitions[begin:end] {
+			topicAssignments := assignments[memberID]
+			if topicAssignments == nil {
+				topicAssignments = make(map[int32]bool, partitionCount)
+				assignments[memberID] = topicAssignments
+			}
+			assignments[memberID][partition] = true
+		}
+		begin = end
+	}
+	return assignments
```

Comment:
Current partition allocation resembles closely (although not 100% - it does not take in account distribution of topics between Kafka brokers) the standard Java Kafka high-level consumer rebalancing [algorithm](http://kafka.apache.org/documentation.html#distributionimpl) (scroll down to **Consumer registration algorithm**). So let's keep it as is. I will add this clarification to the function comment.